test/unit/jwtAuthenicationTest.js

```diff
@@ -0,0 +1,663 @@
+      const ctx = {
+        authenticated: null,
+        request: {
+          header: {
+            authorization: null,
+          },
+        },
+      }
+      const next = sandbox.spy()
+
+      const loggerStub = sandbox.stub(logger, 'warn')
+
+      await jwtAuthentication.koaMiddleware(ctx, next)
+
+      next.callCount.should.eql(1)
+      should(ctx.authenticated).be.null()
+      loggerStub.callCount.should.eql(1)
+    })
+
+    it('should fail due to unknown JWT encryption algorithm', async () => {
+      const ctx = {
+        authenticated: null,
+        request: {
+          header: {
+            authorization:
+              'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0IiwiYXVkIjpbInRlc3RBdWRpZW5jZSJdLCJpc3MiOiJ0ZXN0SXNzdWVyIn0._bUjvzaXWkyYlxV81lVU1dsbZpH_jlW7sda7zsnORwg',
+          },
+        },
+      }
+      const next = sandbox.spy()
+      const loggerStub = sandbox.stub(logger, 'error')
+      const mockJwtConfig = {
+        algorithms: 'INVALID',
+        secretOrPublicKey: 'test',
+      }
+      const configStub = sandbox
+        .stub(configIndex.config, 'get')
+        .onFirstCall()
+        .returns(mockJwtConfig.secretOrPublicKey)
+        .onSecondCall()
+        .returns(mockJwtConfig)
+
+      await jwtAuthentication.koaMiddleware(ctx, next)
+
+      next.callCount.should.eql(1)
+      configStub.callCount.should.eql(2)
+      loggerStub.callCount.should.eql(1)
+      should(ctx.authenticated).be.null()
+    })
+
+    it('should fail due to missing subject field in JWT', async () => {
+      const ctx = {
+        authenticated: null,
+        request: {
+          header: {
+            authorization:
+              'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.P4Lqll22jQQJ1eMJikvNg5HKG-cKB0hUZA9BZFIG7Jk',
+          },
+        },
+      }
+      const next = sandbox.spy()
+      const loggerStub = sandbox.stub(logger, 'error')
+
+      const mockJwtConfig = {
+        algorithms: 'HS256',
+        secretOrPublicKey: 'test',
+      }
+      const configStub = sandbox
+        .stub(configIndex.config, 'get')
+        .onFirstCall()
+        .returns(mockJwtConfig.secretOrPublicKey)
+        .onSecondCall()
+        .returns(mockJwtConfig)
+
+      await jwtAuthentication.koaMiddleware(ctx, next)
+
+      next.callCount.should.eql(1)
+      configStub.callCount.should.eql(2)
+      loggerStub.callCount.should.eql(1)
+      should(ctx.authenticated).be.null()
+    })
+
+    it('should fail due to incorrect secret in config', async () => {
+      const ctx = {
+        authenticated: null,
+        request: {
+          header: {
+            authorization:
+              'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.P4Lqll22jQQJ1eMJikvNg5HKG-cKB0hUZA9BZFIG7Jk',
+          },
+        },
+      }
+      const next = sandbox.spy()
+      const loggerStub = sandbox.stub(logger, 'error')
```

Comment:
wondering if this shouldnt rather be a warning instead of an error. It could just mean that the secret is infact correct but the token used the wrong secret. This would make the request invalid as its not correctly authenticated

Reply:
Is this still valid?

Reply:
Im not entirely sure :/ to me this doesn't feel like an error as its not an exception of any kind, but rather an invalid token. If all 3 mechanisms are enabled, we will get two errors thrown every time a request comes through

Reply:
Lets keep it as is for now and make this change in future if it becomes an issue